File: app/lib/models/member.dart
/// 멤버 모델
library;

class Member {
  final int id;
  final String name;
  final String? nameEn;
  final String? imageUrl;
  final String? imageMedium;
  final String? imageThumb;
  final String? birthDate;
  final String? position;
  final String? instagram;
  final bool isFormer;

  Member({
    required this.id,
    required this.name,
    this.nameEn,
    this.imageUrl,
    this.imageMedium,
    this.imageThumb,
    this.birthDate,
    this.position,
    this.instagram,
    this.isFormer = false,
  });

  factory Member.fromJson(Map<String, dynamic> json) {
    return Member(
      id: json['id'] as int,
      name: json['name'] as String,
      nameEn: json['name_en'] as String?,
      imageUrl: json['image_url'] as String?,
      imageMedium: json['image_medium'] as String?,
      imageThumb: json['image_thumb'] as String?,
      birthDate: json['birth_date'] as String?,
      position: json['position'] as String?,
      instagram: json['instagram'] as String?,
      isFormer: json['is_former'] == 1 || json['is_former'] == true,
    );
  }
}

/// 멤버 컨셉 포토 (개인·유닛 태깅 + 단체)
class MemberPhoto {
  final int id;
  final String? mediumUrl;
  final String? thumbUrl;
  final String? conceptName;
  final String photoType; // 'group' | 'unit' | 'solo'
  final int? width;
  final int? height;
  final String? albumTitle;

  MemberPhoto({
    required this.id,
    this.mediumUrl,
    this.thumbUrl,
    this.conceptName,
    this.photoType = 'group',
    this.width,
    this.height,
    this.albumTitle,
  });

  factory MemberPhoto.fromJson(Map<String, dynamic> json) {
    return MemberPhoto(
      id: json['id'] as int,
      mediumUrl: json['medium_url'] as String?,
      thumbUrl: json['thumb_url'] as String?,
      conceptName: json['concept_name'] as String?,
      photoType: (json['photo_type'] as String?) ?? 'group',
      width: json['width'] as int?,
      height: json['height'] as int?,
      albumTitle: json['album_title'] as String?,
    );
  }
}

File: backend/src/services/member.js
/**
 * 멤버 서비스
 * 멤버 관련 비즈니스 로직
 */
import { getOrSet, invalidate, cacheKeys } from '../utils/cache.js';

/**
 * 전체 멤버 목록 조회 (별명 포함, 캐시 적용)
 * @param {object} db - 데이터베이스 연결
 * @param {object} redis - Redis 클라이언트 (캐시용, 선택적)
 * @returns {array} 멤버 목록
 */
export async function getAllMembers(db, redis = null) {
  const fetchMembers = async () => {
    const [members] = await db.query(`
      SELECT
        m.id, m.name, m.name_en, m.birth_date, m.instagram, m.image_id, m.is_former,
        i.original_url as image_original,
        i.medium_url as image_medium,
        i.thumb_url as image_thumb
      FROM members m
      LEFT JOIN images i ON m.image_id = i.id
      ORDER BY m.is_former ASC, m.id ASC
    `);

    // 별명 조회
    const [nicknames] = await db.query(
      'SELECT member_id, nickname FROM member_nicknames'
    );

    // 멤버별 별명 매핑
    const nicknameMap = {};
    for (const n of nicknames) {
      if (!nicknameMap[n.member_id]) {
        nicknameMap[n.member_id] = [];
      }
      nicknameMap[n.member_id].push(n.nickname);
    }

    // 멤버 데이터에 별명 추가
    return members.map(m => ({
      ...m,
      nicknames: nicknameMap[m.id] || [],
      image_url: m.image_thumb || m.image_medium || m.image_original,
    }));
  };

  // Redis가 있으면 캐시 사용
  if (redis) {
    return getOrSet(redis, cacheKeys.members, fetchMembers, 600); // 10분 캐시
  }
  return fetchMembers();
}

/**
 * 멤버 캐시 무효화
 * @param {object} redis - Redis 클라이언트
 */
export async function invalidateMemberCache(redis) {
  await invalidate(redis, cacheKeys.members);
}

/**
 * 이름으로 멤버 조회 (별명 포함)
 * 한글명(name) 또는 영문명(name_en) 모두 검색 가능
 * @param {object} db - 데이터베이스 연결
 * @param {string} name - 멤버 이름 (한글 또는 영문)
 * @returns {object|null} 멤버 정보 또는 null
 */
export async function getMemberByName(db, name) {
  const [members] = await db.query(`
    SELECT
      m.id, m.name, m.name_en, m.birth_date, m.instagram, m.image_id, m.is_former,
      i.original_url as image_original,
      i.medium_url as image_medium,
      i.thumb_url as image_thumb
    FROM members m
    LEFT JOIN images i ON m.image_id = i.id
    WHERE m.name = ? OR LOWER(m.name_en) = LOWER(?)
  `, [name, name]);

  if (members.length === 0) {
    return null;
  }

  const member = members[0];

  // 별명 조회
  const [nicknames] = await db.query(
    'SELECT nickname FROM member_nicknames WHERE member_id = ?',
    [member.id]
  );

  return {
    ...member,
    nicknames: nicknames.map(n => n.nickname),
    image_url: member.image_original || member.image_medium || member.image_thumb,
  };
}

/**
 * ID로 멤버 기본 정보 조회 (수정용)
 * @param {object} db - 데이터베이스 연결
 * @param {string} name - 멤버 이름
 * @returns {object|null} 멤버 기본 정보 또는 null
 */
export async function getMemberBasicByName(db, name) {
  const [members] = await db.query(
    'SELECT id, image_id FROM members WHERE name = ?',
    [name]
  );
  return members.length > 0 ? members[0] : null;
}
